drivers/SmartThings/philips-hue/src/disco/contact.lua

```diff
@@ -0,0 +1,127 @@
+    rid_by_rtype[svc.rtype] = svc.rid
+  end
+
+  local contact, contact_err = api_instance:get_contact_by_id(rid_by_rtype[HueDeviceTypes.CONTACT])
+  if contact_err then return nil, contact_err end
+
+  local tamper, tamper_err = api_instance:get_tamper_by_id(rid_by_rtype[HueDeviceTypes.TAMPER])
+  if tamper_err then return nil, tamper_err end
+
+  local battery, battery_err = api_instance:get_device_power_by_id(rid_by_rtype[HueDeviceTypes.DEVICE_POWER])
+  if battery_err then return nil, battery_err end
+
+  local resource_id = rid_by_rtype[HueDeviceTypes.CONTACT]
+  local contact_sensor_description = {
+    hue_provided_name = device_service_info.metadata.name,
+    id = resource_id,
+    parent_device_id = bridge_id,
+    hue_device_id = device_service_info.id,
+    hue_device_data = device_service_info,
+  }
+
+  if contact and contact.data and contact.data[1] then
+    contact_sensor_description.contact_report = contact.data[1].contact_report
+    contact_sensor_description.contact_enabled = contact.data[1].enabled
+
+  end
+
+  if battery and battery.data and battery.data[1] then
+    contact_sensor_description.power_id = battery.data[1].id
+    contact_sensor_description.power_state = battery.data[1].power_state
+  end
+
+  if tamper and tamper.data and tamper.data[1] then
+    contact_sensor_description.tamper_id = tamper.data[1].id
+    contact_sensor_description.tamper_reports = tamper.data[1].tamper_reports
+  end
+
+  if type(cache) == "table" then
+    cache[resource_id] = contact_sensor_description
+    if device_service_info.id_v1 then
+      cache[device_service_info.id_v1] = contact_sensor_description
+    end
+  end
+
+  return contact_sensor_description
+end
+
+---@param api_instance PhilipsHueApi
+---@param device_service_id string
+---@param bridge_id string
+---@param cache table?
+---@return table<string,any>? description nil on error
+---@return string? err nil on success
+function M.update_state_for_all_device_services(api_instance, device_service_id, bridge_id, cache)
+  log.debug("----------- Calling REST API")
+  local device_service_info, err = api_instance:get_device_by_id(device_service_id)
+  if err or not (device_service_info and device_service_info.data) then
+    log.error("Couldn't get device info for sensor, error: " .. st_utils.stringify_table(err))
+    return
+  end
+
+  log.debug("------------ _do_update")
+  return _do_update(api_instance, device_service_info.data[1], bridge_id, cache)
+end
+
+---@param driver HueDriver
+---@param bridge_id string
+---@param api_instance PhilipsHueApi
+---@param resource_id string
+---@param device_service_info HueDeviceInfo
+---@param device_state_disco_cache table<string, table>
+---@param st_metadata_callback fun(driver: HueDriver, metadata: table)?
+function M.handle_discovered_device(
+    driver, bridge_id, api_instance,
+    resource_id, device_service_info,
+    device_state_disco_cache, st_metadata_callback
+)
+  local err = select(2,
```

Comment:
What is this select doing?

Reply:
That's a native Lua select, not an async select.

It throws away all arguments that come before the given index in a variadic return situation. The benefit here compared to binding to `_` is that you don't take up memory with a variable binding. It's basically saying we don't care about the first return value, we only care if there's an error.